### ISSUE TITLE
Security scan reported problem with com.google.common.io.FileBackedOutputStream

### ISSUE THREAD

**Issue by venusjain10** [google/guava#2575]
We are using google commons APIs in our project and security scan on our code has reported following problem. Could you please let me know if this is of any concern or it can be fixed in new version of this API.

**Problem:** Creating and using insecure temporary files can leave application and system data vulnerable to attack. In particular,file names created by the tmpnam family of functions can be easily guessed by an attacker. If an attacker can predictthe filename and create a malicious collision, he may be able to manipulate the behavior of the application.

**Recommendations (Generated by tool)**
Ensure that unpredictable names are used for temporary files and that files are created in a secure directory with appropriate permissions. Using mkstemp() is a reasonably safe way to create temporary files. It will attempt to create and open a unique file based on a filename template provided by the user, combined with a series of randomly generated characters. Note that mkstemp() is safe if only the descriptor is used and the returned filename is not used in a subsequent function call with extra privileges. Using mkstemp() does not completely eliminate race conditions but does provide better protection than other methods

Class:
com.google.common.io.FileBackedOutputStream

**Comment by lowasser** [comment]
This seems like it could at least conceivably be a problem if, and only if, you used FileBackedOutputStream.  Do you?

**Comment by abhisheksharma43** [comment]
I have not used the class FileBackedOutputStream directly. However I am not sure if the classes that we have used from the API internally use FileBackedOutputStream.

**Comment by cpovirk** [comment]
I have a request in for a CVE for this, which is fundamentally the same as #4011. The just-merged fix will be part of the next release, hopefully to be published tomorrow.

**Comment by cpovirk** [comment]
(We have the number CVE-2023-2976 reserved, but it's not posted yet.)

**Comment by somera** [comment]
Hi, I have the finding for 32.0.0 too:
`[ERROR] Failed to execute goal org.owasp:dependency-check-maven:8.2.1:aggregate (default-cli) on project brm-parent: 
[ERROR]
[ERROR] One or more dependencies were identified with vulnerabilities that have a CVSS score greater than or equal to '4,0': 
[ERROR]
[ERROR] guava-32.0.0-jre.jar: CVE-2023-2976(6.2)
[ERROR]
[ERROR] See the dependency-check report for more details.`

**Comment by Horcrux7** [comment]
The CVE-2023-2976 match also to the version 32.0.0. Is it fixed and the CVE version match is wrong or will it fix in the next version?

**Comment by pwagland** [comment]
SonaType also flags this CVE as being present in 32.0.0, despite [the release notes](https://github.com/google/guava/releases/tag/v32.0.0) saying the release fixes CVE-2023-2976.

Basically what do we need to do:
1. Wait for upstream to fix their data, or
2. Suppress this CVE, since we don't use these classes _anyway_, or
3. Wait for a new fix, aka 32.0.1?

Trying not to just say "me too", I hope this is additional useful information.

**Comment by malfoj89** [comment]
yesteday it was okay - today it matches 32.0.0 any comment on that?

**Comment by cpovirk** [comment]
Thanks for the reports. I'd welcome any help in figuring this out on https://github.com/google/guava/issues/6532.

**Comment by cpovirk** [comment]
For the record, the CVE is now published: https://www.cve.org/CVERecord?id=CVE-2023-2976

**Comment by cpovirk** [comment]
You might be interested in the release notes [for 31.0](https://github.com/google/guava/releases/tag/v31.0) and [for 32.0.0](https://github.com/google/guava/releases/tag/v32.0.0). Even though both releases are major-version bumps, they maintain binary compatibility for the Guava jar. They were major-version bumps primarily because we removed some methods from the GWT jar that we provide, which doesn't affect normal users.

**Comment by cpovirk** [comment]
([Our policy](https://github.com/google/guava#important-warnings) (in effect [since 2017](https://groups.google.com/g/guava-announce/c/6mtQTDDADsg)) does still allow us to remove _`@Beta`_ APIs. But after some removals in 2018, I believe we removed [two](https://github.com/google/guava/commit/3dd22fe75a1dce9dcb169a91e3b617da5bb77587) [things](https://github.com/google/guava/commit/dfb00017144bf48c408bc4cfcebadc0a0498dc73) in the next couple years and then nothing in the almost 3 years since. Plus, we've removed `@Beta` from almost all our APIs, so there aren't many APIs left that we could remove :)

**Comment by jbduncan** [comment]
@cpovirk Good point, I'd forgotten about the `@Beta` APIs. But given there's so few of them now compared to a few years ago, I think @sourabhsparkala's migration will be less painful that it would have been. :)

**Comment by feilimb** [comment]
Can I ask if Windows 10 users are affected by this vulnerability, I am not entirely clear from reading of this issue whether this is related to Unix / Android OS only, or whether Windows users are also affected.

**Comment by cpovirk** [comment]
First, I should emphasize that almost no app uses `FileBackedOutputStream`. And if an app doesn't use it, then it's not affected.

If an app _does_ use it, then my understanding is that the app is still safe _under Windows_ because Windows gives each user a private temporary directory:

- https://github.com/google/guava/issues/6535#issuecomment-1577022231
- https://github.com/google/guava/issues/6535#issuecomment-1579806211

So, as best I understand, the problem is restricted to the cases laid out in CVE-2023-2976:
- Linux systems whose temporary directory might be accessible to people you don't want to see your `FileBackedOutputStream` data. We made this case secure in new Guava versions.
- Android Ice Cream Sandwich. That version is so old that probably no one is it, so we broke `FileBackedOutputStream` entirely in new Guava versions.
